File: opinionpulse/opinionpulse-frontend/src/api/billing.ts
import { apiRequest } from "@/api/client"
import type { PlanId } from "@/data/pricingData"

export type BillingInterval = "monthly" | "annual"

export type BillingConfig = {
  configured: boolean
  publishable_key: string | null
}

export type CheckoutUiMode = "hosted" | "embedded"

export type CheckoutSessionResponse = {
  checkout_url?: string | null
  client_secret?: string | null
  session_id: string
}

export type PortalSessionResponse = {
  portal_url: string
}

export type CheckoutSessionStatus = {
  status: string
  plan_id: string | null
  plan_name: string | null
  synced: boolean
}

export function getBillingConfig() {
  return apiRequest<BillingConfig>("/api/billing/config")
}

export function createCheckoutSession(
  planId: PlanId,
  interval: BillingInterval,
  uiMode: CheckoutUiMode = "hosted"
) {
  return apiRequest<CheckoutSessionResponse>("/api/billing/checkout-session", {
    method: "POST",
    auth: true,
    body: { plan_id: planId, interval, ui_mode: uiMode },
  })
}

export function createPortalSession() {
  return apiRequest<PortalSessionResponse>("/api/billing/portal-session", {
    method: "POST",
    auth: true,
  })
}

export function syncCheckoutSession(sessionId: string) {
  return apiRequest<CheckoutSessionStatus>(`/api/billing/session/${sessionId}`, {
    auth: true,
  })
}

File: opinionpulse/opinionpulse-backend/app/schemas/billing.py
from typing import Literal, Optional

from pydantic import BaseModel

BillingInterval = Literal["monthly", "annual"]
PaidPlanId = Literal["starter", "pro", "enterprise"]
CheckoutUiMode = Literal["hosted", "embedded"]


class CheckoutSessionRequest(BaseModel):
    plan_id: PaidPlanId
    interval: BillingInterval = "monthly"
    ui_mode: CheckoutUiMode = "hosted"


class CheckoutSessionResponse(BaseModel):
    session_id: str
    checkout_url: Optional[str] = None
    client_secret: Optional[str] = None


class PortalSessionResponse(BaseModel):
    portal_url: str


class BillingConfigResponse(BaseModel):
    configured: bool
    publishable_key: Optional[str] = None


class CheckoutSessionStatusResponse(BaseModel):
    status: str
    plan_id: Optional[str] = None
    plan_name: Optional[str] = None
    synced: bool = False
